src/server.ts

```diff
@@ -699,6 +724,68 @@ export class Server {
       offset,
     );
   }
+
+  /**
+   * Check if any of the destination accounts requires a memo.
+   *
+   * This function implements a memo required check as defined in
+   * [SEP0029](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0029.md).
+   * It will load each account which is the destination and check if it has the
+   * data field `config.memo_required` set to `"MQ=="`.
+   *
+   * @see
+   * [SEP0029](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0029.md)
+   * @param {Transaction} transaction - The transaction to check.
+   * @returns {Promise<void, Error>} - If any of the destination account
+   * requires a memo, the promise will throw {@link AccountRequiresMemoError}.
+   * @throws  {AccountRequiresMemoError}
+   */
+  public async checkMemoRequired(transaction: Transaction): Promise<void> {
+    const destinations = new Set<string>();
+
+    for (let i = 0; i < transaction.operations.length; i++) {
+      const operation = transaction.operations[i];
+
+      switch (operation.type) {
+        case "payment":
+        case "pathPaymentStrictReceive":
+        case "pathPaymentStrictSend":
+        case "accountMerge":
+          break;
+        default:
+          continue;
+      }
+      const destination = operation.destination;
+      if (destinations.has(destination)) {
+        continue;
+      }
+      destinations.add(destination);
+
+      try {
+        const account = await this.loadAccount(destination);
```

Comment:
This `await` inside a `for` is gonna load accounts serially, might be worth restructuring. It looks like this could be a big slowdown if there's a single transaction sending to a lot of unique destinations. Maybe a map/filter or reduce down to a unique list of destinations, then `Promise.all` and load all accounts in one go?

Reply:
@vcarl for this first version we want to load accounts serially -- as for the scenario you described, that's why we have this check https://github.com/stellar/js-stellar-sdk/pull/516/files#diff-916156c481141bfcd08c6ae40ce10e00R759 -- if the account was already checked, the we skip it.

Reply:
I like that suggestion too.
We could have 100 operations inside a transaction, which means this validation could possibly take more than a minute.

Reply:
@marcelosalloum @vcarl hey we are aware of this. For this version we decided to go ahead and it sequentially even thought it will mean 100 calls (in the worst case) -- you can read more about it https://github.com/stellar/go/issues/2329#issuecomment-597242134 

> We will not implement caching or potentially batching memo checks at this time. Both add significant complexity. User can avoid the need for time-based caching if they control the account and track setting the memo id. Batching needs significant thought as i) it may eventually span other endpoints; ii) it adds a lot of complexity to evaluating received responses; iii) database implications need evaluation.

Reply:
Gotcha! Would you consider making this potential issue more explicit in the `submitTransaction` method documentation?

Reply:
@marcelosalloum I added a note on `checkMemoRequired` and I'll call it out on the changelog.

Reply:
It seems like those requests could happen in parallel though. Not batching, just not with each request waiting for the previous request to complete. The code right now will make a network graph that looks like
```
|---|
    |---|
        |---|
```

Reply:
Even if we decided to defer this for the Go SDK, where doing so might be significantly complicated, it might make sense to implement it straight up here if doing so is not significantly complicated. @vcarl @abuiles Is this significantly complicated to implement?

Reply:
We discussed this on slack, the code is doing what we are expecting, it will load each account one by one -- you can checkout an example here https://repl.it/repls/DiscreteToughSearchengine 

We explicitly want to load each account instead of loading multiple at the same time.

Reply:
@leighmcculloch  the reason why we are doing this is not because the code is complicated but rather we wanted to avoid kicking rate limiting in horizon. For this first iteration we were thinking to release like this and see what feedback did we get. Worse case scenario people can opt-out of the check.